thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp

```diff
@@ -45,1910 +49,2237 @@
+    ISerialStream *stream = nullptr;
+};
+
+class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor
+{
+    typedef CSlaveActivity PARENT;
+
+    class CLimiter
     {
-        CriticalBlock c(crit);
-        if (preserveOrder)
+        unsigned max, leeway;
+        Semaphore sem;
+        CriticalSection crit;
+        unsigned count = 0;
+        unsigned blocked = 0;
+    public:
+        CLimiter() : max(0), leeway(0)
+        {
+        }
+        CLimiter(unsigned _max, unsigned _leeway=0) : max(_max), leeway(_leeway)
+        {
+        }
+        void set(unsigned _max, unsigned _leeway=0)
+        {
+            max = _max;
+            leeway = _leeway;
+        }
+        bool preIncNonBlocking()
         {
-            CJoinGroup *finger = head.next;
-            if (preserveGroups)
             {
-                unsigned cnt=0;
-                while (finger != &head)
-                {
-                    if (finger->complete())
-                    {
-                        CJoinGroup *next = processRemoveGroup(finger);
-                        notify.addJoinGroup(finger); // means doneQueue owns linked list
-                        finger = next;
-                        ++cnt;
-                    }
-                    else
-                        break;
-                }
+                CriticalBlock b(crit);
+                if (count++ < max+leeway)
+                    return false;
+                ++blocked;
             }
-            else
+            return true;
+        }
+        bool incNonBlocking()
+        {
             {
-                while (finger != &head)
+                CriticalBlock b(crit);
+                if (count < max+leeway)
                 {
-                    if (finger->complete())
-                    {
-                        CJoinGroup *next = finger->next;
-                        removeJoinGroup(finger);
-                        notify.addJoinGroup(finger);
-                        finger = next;
-                    }
-                    else
-                        break;
+                    ++count;
+                    return false;
                 }
+                ++blocked;
             }
+            return true;
         }
-        else if (preserveGroups)
-        {
-            CJoinGroup *next = processRemoveGroup(jg);
-            notify.addJoinGroup(jg); // means doneQueue owns linked list
-            assertex(!next->inGroup(jg));
-        }
-        else
+        void inc()
         {
-            removeJoinGroup(jg);
-            notify.addJoinGroup(jg);
+            while (incNonBlocking())
+                sem.wait();
         }
-    }
-#ifdef TRACE_USAGE
-    void getStats(StringBuffer &str)
-    {
-        UnsignedArray counts;
+        void dec()
         {
-            CriticalBlock c(crit);
-            CJoinGroup *p = &head;
-            for (;;)
+            CriticalBlock b(crit);
+            --count;
```

Comment:
@ghalliday - will come back to this (and 1 other comment), am going to make rest of changes, test and push 1st.

Reply:
I think you are correct that it will not cause problems in practice.

"I think it means that if max is 10, preinc() is called 10 times, then 10 times more (same threads).  None of those threads will l be allowed to run until the first 10 rows have all been processed.  That probably isn't a disaster.

>the threads use cases don't call preInc
"

Reply:
Yes, preIncNonBlocking() is intended to be used / is used for other cases, not for counting running threads. It's used when there's a limiter on rows in queues, where the row is always added, the limit triggers blocking, but other threads are independently pulling (and decrementing).